### ISSUE TITLE
Consistency with sorting: `sort=True`

### ISSUE THREAD

**Issue by Deepblue129** [pytorch/text#69]
**Problem:**
```
    train_iter, dev_iter, test_iter = data.BucketIterator.splits(
        (train, dev, test),
        batch_sizes=(32, 256, 256),
        sort_key=lambda x: len(x.input),
        sort=True,
        device=-1)  # Use CPU
```

If `sort=True` and `train=True`, then the `train_iter` batches are shuffled. This behavior is unexpected.

**Cause:**
Because by default `self.shuffle=True` is `train=True`. Then https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L113 `shuffle` overrides `sort`.

**Possible Solution:**
`sort=True` should override `shuffle=None and train=True`.

**Comment by jekbradbury** [comment]
I think you're right, but I'll look into it more on Monday (a couple times a seemingly obvious, one-line change like this has broken existing code and I want to make sure that's unlikely here).
